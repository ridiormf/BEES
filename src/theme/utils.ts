type Space = number | string;

const utils = {
  marginHorizontal: (value: Space) => ({
    marginLeft: value,
    marginRight: value,
  }),
  marginVertical: (value: Space) => ({
    marginTop: value,
    marginBottom: value,
  }),
  paddingHorizontal: (value: Space) => ({
    paddingLeft: value,
    paddingRight: value,
  }),
  paddingVertical: (value: Space) => ({
    paddingTop: value,
    paddingBottom: value,
  }),
};

export default utils;
