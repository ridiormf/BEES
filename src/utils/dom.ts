export const getBody = () => {
  return document.getElementsByTagName('body')[0];
};
