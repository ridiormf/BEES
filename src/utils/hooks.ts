import React from 'react';
import { useLocation, useNavigate } from 'react-router-dom';

/**
 * Function that simulates the behavior of a class component constructor,
 * adapted to hooks, for functional components.
 *
 * @param {function} callback Callback which will be called as soon as the component is instantiated.
 */
export const useConstructor = <ValueType = void>(
  callback: () => ValueType,
): ValueType | void => {
  const componentWasBuilt = React.useRef<boolean>(false);
  const value = React.useRef<ValueType>();
  if (!componentWasBuilt.current) {
    componentWasBuilt.current = true;
    value.current = callback();
  }

  return value.current;
};

/**
 * Function that simulates the behavior of a class component method "componentDidMount",
 * adapted to hooks, for functional components.
 *
 * @param {function} callback Callback which will be called as soon as the component mount.
 */
export const useDidMount = (callback: () => void): void => {
  React.useEffect(callback, []);
};

/**
 * Function that simulates the behavior of a class component method "componentWillUnmount",
 * adapted to hooks, for functional components.
 *
 * @param {function} callback Callback which will be called as soon as the component will unmount.
 */
export const useWillUnmount = (callback: () => void): void => {
  React.useEffect(() => {
    return callback;
  }, []);
};

/**
 * Function that simulates the behavior of a class component method "componentDidUpdate",
 * adapted to hooks, for functional components.
 *
 * @param {function} callback Callback which will be called on each value change of variables declared as dependencies,
 * excluding the first render of the component.
 * @param {React.DependencyList} deps Dependency list to be listened.
 */
export const useDidUpdate = (
  callback: () => void,
  deps: React.DependencyList,
): void => {
  const firstRender = React.useRef<boolean>(true);
  React.useEffect(() => {
    if (firstRender.current) {
      firstRender.current = false;
    } else {
      callback();
    }
  }, deps);
};

/**
 * Function that use main hooks from "react-router-dom" to simplify usability.
 *
 * @returns {object} containing following attributes:
 * @type {function} "goBack": to return navigation one page.
 * @type {function} "go": to navigate for path.
 * @type {string} "pathname": current url pathname.
 */
export const useAppNavigation = () => {
  const navigate = useNavigate();
  const location = useLocation();

  const goBack = () => {
    navigate(-1);
  };

  const go = (path: string) => {
    navigate(path);
  };

  return {
    goBack,
    go,
    pathname: location.pathname,
  };
};
