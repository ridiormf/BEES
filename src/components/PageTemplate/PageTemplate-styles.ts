import { CSS } from '@stitches/react';
import theme from '../../theme';

const Container = theme.styled('div', {
  alignSelf: 'stretch',
  flexGrow: 1,
  backgroundColor: '$secondaryBackground',
  display: 'flex',
  flexDirection: 'column',
});

const Header = theme.styled('header', {
  height: 72,
  backgroundColor: '$primaryBackground',
  color: '$primaryText',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'space-between',
  paddingHorizontal: 20,
});

const BackButton = theme.styled('button', {
  display: 'flex',
  alignItems: 'center',
  padding: 0,
  lineHeight: 0,
  border: 'none',
  backgroundColor: 'unset',
  cursor: 'pointer',
  transition: '200ms',
  '&:hover': {
    filter: 'brightness(1.5)',
  },
  '&:active': {
    filter: 'brightness(2)',
  },
});

const HeaderTitles = theme.styled('h2', {
  fontSize: 30,
  fontWeight: 400,
  marginLeft: 16,
  textTransform: 'capitalize',
  '@smDown': {
    fontSize: 20,
  },
});

const ContentContainer = theme.styled('div', {
  padding: 60,
  display: 'flex',
  flexDirection: 'column',
  flexGrow: 1,
  '@lgDown': {
    padding: 40,
  },
  '@smDown': {
    padding: 20,
  },
});

const backImageCSS: CSS = {};

export const pageTemplateStyles = {
  styled: {
    Container,
    Header,
    BackButton,
    HeaderTitles,
    ContentContainer,
  },
  css: {
    backImage: backImageCSS,
  },
};
