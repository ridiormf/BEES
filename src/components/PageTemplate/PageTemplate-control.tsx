import { useApplicationContext } from '../../providers/ApplicationProvider';
import { useBreweriesContext } from '../../providers/BreweriesProvider';
import { useAppNavigation } from '../../utils/hooks';

export const usePageTemplateControl = () => {
  const { user, clearUser } = useApplicationContext();
  const { clearBreweries } = useBreweriesContext();
  const { goBack, pathname } = useAppNavigation();

  const logout = () => {
    clearUser!();
    clearBreweries!();
  };

  const onGoBackClick = () => {
    if (pathname === '/') {
      logout();
    } else {
      goBack();
    }
  };

  const fullName = user?.fullName;
  return {
    state: {
      fullName,
    },
    methods: {
      onGoBackClick,
    },
  };
};
