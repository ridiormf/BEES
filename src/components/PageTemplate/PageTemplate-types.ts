import React from 'react';
import { pageTemplateStyles } from './PageTemplate-styles';

const {
  styled: { ContentContainer },
} = pageTemplateStyles;

export interface PageTemplateProps
  extends React.ComponentProps<typeof ContentContainer> {
  children: React.ReactNode | React.ReactNode[];
}
