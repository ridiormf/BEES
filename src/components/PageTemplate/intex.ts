import PageTemplate from './PageTemplate';
import type { PageTemplateProps } from './PageTemplate-types';

export default PageTemplate;
export type { PageTemplateProps };
