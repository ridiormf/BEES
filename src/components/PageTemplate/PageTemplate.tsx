import React, { ReactComponentElement } from 'react';
import { usePageTemplateControl } from './PageTemplate-control';

import { pageTemplateStyles } from './PageTemplate-styles';
import { PageTemplateProps } from './PageTemplate-types';

import backButtonSvg from '../../assets/svg/back.svg';
import Image from '../Image';

const {
  styled: { Container, Header, BackButton, HeaderTitles, ContentContainer },
} = pageTemplateStyles;

const PageTemplate: React.FC<PageTemplateProps> = ({ children }) => {
  const {
    state: { fullName },
    methods: { logout },
  } = usePageTemplateControl();
  return (
    <Container>
      <Header>
        <BackButton onClick={logout}>
          <Image src={backButtonSvg} />
          <HeaderTitles>Go Back</HeaderTitles>
        </BackButton>
        <HeaderTitles>{fullName}</HeaderTitles>
      </Header>
      <ContentContainer>{children}</ContentContainer>
    </Container>
  );
};

export default PageTemplate;
