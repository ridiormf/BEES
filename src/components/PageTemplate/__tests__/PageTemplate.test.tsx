import { render } from '@testing-library/react';
import { Wrapper } from '../../../__tests__/utils';
import PageTemplate from '../PageTemplate';

const PATHNAME = '/test-pathname';

jest.mock('react-router-dom', () => ({
  ...(jest.requireActual('react-router-dom') as any),
  useNavigate: jest.fn(),
  useLocation: () => ({ pathname: PATHNAME }),
}));

describe('components ->', () => {
  describe('PageTemplate ->', () => {
    describe('snapshots ->', () => {
      it('Should match snapshot', () => {
        const { container } = render(
          <PageTemplate>
            <></>
          </PageTemplate>,
          {
            wrapper: Wrapper,
          },
        );
        expect(container).toMatchSnapshot();
      });
    });
  });
});
