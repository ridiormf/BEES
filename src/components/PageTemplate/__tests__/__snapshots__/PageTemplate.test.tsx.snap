// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components -> PageTemplate -> snapshots -> Should match snapshot 1`] = `
<div>
  <div
    class="c-jAbhDY"
  >
    <header
      class="c-jveizM"
    >
      <button
        class="c-dWdZEg"
      >
        <img
          alt="go back icon"
          class="PJLV PJLV-iPJLV-css"
          src="arrow-circle-left.svg"
        />
        <h2
          class="c-kWbkwW"
        >
          Go Back
        </h2>
      </button>
      <h2
        class="c-kWbkwW"
      />
    </header>
    <div
      class="c-bmCckb"
    />
  </div>
</div>
`;
